src/soc/sifive/fu540/src/clock.rs

```diff
@@ -376,59 +275,37 @@ impl Clock {
         // get DDR out of reset
         // TODO: clean this up later
         self.DevReset.write(ResetCtl::DDRCtl::Reset);
-        //write32(&prci->devicesresetreg, PRCI_DEVICESRESET_DDR_CTRL_RST_N(1));
 
-        // TODO need a fence MACRO. Can't be a call to a function.
-        // 	// HACK to get the '1 full controller clock cycle'.
         architecture::fence();
-        // 	asm volatile ("fence");
-        // YUCK.
+
         self.DevReset.set(ResetMask(false, false, false, false, true));
-        // 	// get all DDR out of reset
-        // 	write32(&prci->devicesresetreg,
-        // 		PRCI_DEVICESRESET_DDR_CTRL_RST_N(1) |
-        // 		PRCI_DEVICESRESET_DDR_AXI_RST_N(1) |
-        // 		PRCI_DEVICESRESET_DDR_AHB_RST_N(1) |
-        // 		PRCI_DEVICESRESET_DDR_PHY_RST_N(1));
-
-        // 	// HACK to get the '1 full controller clock cycle'.
+
+        // HACK to get the '1 full controller clock cycle'.
         architecture::fence();
-        // 	asm volatile ("fence");
 
-        // 	// These take like 16 cycles to actually propagate. We can't go sending
-        // 	// stuff before they come out of reset. So wait.
-        // 	// TODO: Add a register to read the current reset states, or DDR Control
-        // 	// device?
+        // These take like 16 cycles to actually propagate. We can't go sending
+        // stuff before they come out of reset. So wait.
+        // TODO: Add a register to read the current reset states, or DDR Control
+        // device?
         for i in 0..255 {
             architecture::nop();
         }
-        // 	for (int i = 0; i < 256; i++)
-        // 		asm volatile ("nop");
-        // FIXME need a delay
-
         self.init_pll_ge();
         self.DevReset.set(ResetMask(false, false, false, false, false));
-        // 	write32(&prci->devicesresetreg,
-        // 		PRCI_DEVICESRESET_DDR_CTRL_RST_N(1) |
-        // 		PRCI_DEVICESRESET_DDR_AXI_RST_N(1) |
-        // 		PRCI_DEVICESRESET_DDR_AHB_RST_N(1) |
-        // 		PRCI_DEVICESRESET_DDR_PHY_RST_N(1) |
-        // 		PRCI_DEVICESRESET_GEMGXL_RST_N(1));
-
-        // TODO fence
-        // 	asm volatile ("fence");
+
+        architecture::fence();
     }
 }
 
 // NASTY. We'll live with it for now.
 const UART_DEVICES: u32 = 2;
 const UART_REG_DIV: u32 = 0x18;
-const UART_DIV_VAL: u32 = 4;
+const UART_DIV_VAL: u32 = 4340;
```

Comment:
I wonder if due to other bugs coreboot never got this right ...

Reply:
Yes, 4 is definitely wrong. The datasheet even says the value cannot be less than 16.